zebra-rpc/src/methods/get_block_template_rpcs/get_block_template.rs

```diff
@@ -28,21 +28,70 @@ use crate::methods::get_block_template_rpcs::{
 
 pub use crate::methods::get_block_template_rpcs::types::get_block_template::*;
 
+use super::types::hex_data::HexData;
+
 // - Parameter checks
 
-/// Returns an error if the get block template RPC `parameters` are invalid.
-pub fn check_block_template_parameters(
-    parameters: &get_block_template::JsonParameters,
-) -> Result<()> {
-    if parameters.data.is_some() || parameters.mode == GetBlockTemplateRequestMode::Proposal {
-        return Err(Error {
-            code: ErrorCode::InvalidParams,
-            message: "\"proposal\" mode is currently unsupported by Zebra".to_string(),
-            data: None,
-        });
+impl JsonParameters {
+    /// Checks that `data` is omitted in `Template` mode or provided in `Proposal` mode,
+    ///
+    /// Returns an error if there's a mismatch between the mode and whether `data` is provided.
+    /// Returns Ok(Some(data)) with the block proposal hexdata if in `Proposal` mode.
+    /// Returns Ok(None) if in `Template` mode.
+    pub fn block_proposal(parameters: Option<Self>) -> Result<Option<HexData>> {
```

Comment:
I'm finding this method hard to understand and review.

It seems to do two or three different things:
- In template mode, check that the proposal parameters are valid (that is, both not provided)
- In proposal mode, check that the proposal parameters are valid (that is, both provided)
- In proposal mode, return the proposal parameters

Looking at its name, it is not obvious that it needs to be called in both modes. It looks like a method that gets the data, regardless of any parameter checks.

Optional: We could also split it into separate check and get methods, if you want. And I wonder if taking `&self` would make this method easier to understand and use.

```rust
    /// Checks that `data` is omitted in `Template` mode or provided in `Proposal` mode,
    /// then returns the proposal data.
    ///
    /// Returns an error if there's a mismatch between the mode and whether `data` is provided.
    /// Returns `Ok(Some(hexdata))` with the block proposal `hexdata` if in `Proposal` mode.
    /// Returns `Ok(None)` if in `Template` mode.
    pub fn check_block_proposal(parameters: Option<Self>) -> Result<Option<HexData>> {
```

Reply:
I didn't want to repeat some of the code but it's probably better to split it into a different check if it's confusing. Updated in [`8420b75`](https://github.com/ZcashFoundation/zebra/pull/5870/commits/8420b75013c1a05ea11a9799d1f87bd338c3c27c).